### PR TITLE
pages not restricted for dimagi when flag is enable for domain

### PR DESCRIPTION
**Pull request by lwyszomi**
@czue 
http://manage.dimagi.com/default.asp?227648#1151493

### CODE REVIEW

File: corehq/apps/domain/decorators.py [hunk 1]
```diff
@@ -84,6 +87,9 @@ def _inner(req, domain, *args, **kwargs):
                 else:
                     # some views might not have this set
                     couch_user = CouchUser.from_django_user(user)
+
+                dimagi_pages = [x for x in pages_not_restricted_for_dimagi if x % {'domain': domain_name} == req.path]
```
Comment (dimagimon) [inline corehq/apps/domain/decorators.py]: E501 line too long (118 > 115 characters)

File: corehq/apps/domain/decorators.py [hunk 2]
```diff
@@ -319,7 +327,8 @@ def _inner(request, domain, *args, **kwargs):
             domain_name, domain = load_domain(request, domain)
             if not domain:
                 raise Http404()
-            if not request.couch_user.is_domain_admin(domain_name):
+            dimagi_pages = [x for x in pages_not_restricted_for_dimagi if x % {'domain': domain_name} == request.path]
```
Comment (dimagimon) [inline corehq/apps/domain/decorators.py]: E501 line too long (118 > 115 characters)

File: corehq/apps/domain/decorators.py [hunk 3]
```diff
@@ -84,6 +87,11 @@ def _inner(req, domain, *args, **kwargs):
                 else:
                     # some views might not have this set
                     couch_user = CouchUser.from_django_user(user)
+
+                dimagi_pages = [
+                    x for x in pages_not_restricted_for_dimagi if x % {'domain': domain_name} == req.path
+                ]
```
Comment (czue) [inline corehq/apps/domain/decorators.py]: I think you could combine this addition with the declaration above into a function with signature `_page_is_whitelisted(path, domain)` and then just drop that into the if statements which would probably be easier to read (and could be reused in both places).

File: corehq/apps/domain/decorators.py [hunk 4]
```diff
@@ -94,7 +103,7 @@ def _inner(req, domain, *args, **kwargs):
                     else:
                         return view_func(req, domain_name, *args, **kwargs)
 
-                elif user.is_superuser and not domain.restrict_superusers:
+                elif user.is_superuser and (_page_is_whitelist(req.path, domain_name) or not domain.restrict_superusers):
```
Comment (dimagimon) [inline corehq/apps/domain/decorators.py]: E501 line too long (121 > 115 characters)

File: corehq/apps/domain/decorators.py [hunk 5]
```diff
@@ -94,7 +103,8 @@ def _inner(req, domain, *args, **kwargs):
                     else:
                         return view_func(req, domain_name, *args, **kwargs)
 
-                elif user.is_superuser and not domain.restrict_superusers:
+                elif user.is_superuser and \
+                     (_page_is_whitelist(req.path, domain_name) or not domain.restrict_superusers):
```
Comment (dimagimon) [inline corehq/apps/domain/decorators.py]: E127 continuation line over-indented for visual indent

File: corehq/apps/domain/decorators.py [hunk 6]
```diff
@@ -94,7 +103,10 @@ def _inner(req, domain, *args, **kwargs):
                     else:
                         return view_func(req, domain_name, *args, **kwargs)
 
-                elif user.is_superuser and not domain.restrict_superusers:
+                elif (
+                    _page_is_whitelist(req.path, domain_name) or
+                    not domain.restrict_superusers
+                ) and user.is_superuser :
```
Comment (dimagimon) [inline corehq/apps/domain/decorators.py]: E203 whitespace before ':'

**Comment by lwyszomi** [comment]
@czue 
Updated

File: corehq/apps/domain/decorators.py [hunk 7]
```diff
@@ -319,7 +331,8 @@ def _inner(request, domain, *args, **kwargs):
             domain_name, domain = load_domain(request, domain)
             if not domain:
                 raise Http404()
-            if not request.couch_user.is_domain_admin(domain_name):
+            if not _page_is_whitelist(request.path, domain_name) and \
```
Comment (czue) [inline corehq/apps/domain/decorators.py]: I think this clause also needs a `user.is_superuser` clause otherwise whitelisted pages will be accessible for any user.

File: corehq/apps/domain/decorators.py [hunk 8]
```diff
@@ -59,13 +59,21 @@ def _redirect_for_login_or_domain(request, redirect_field_name, login_url):
     return HttpResponseRedirect(nextURL)
 
 
+def _page_is_whitelist(path, domain):
+    pages_not_restricted_for_dimagi = getattr(settings, "PAGES_NOT_RESTRICTED_FOR_DIMAGI", tuple())
+    return [
+        x for x in pages_not_restricted_for_dimagi if x % {'domain': domain} == path
+    ]
```
Comment (czue) [inline corehq/apps/domain/decorators.py]: nit: might be clear if this returned a `bool()` on the list.

File: corehq/apps/domain/decorators.py (hunk 7 above)
Comment (lwyszomi) [inline corehq/apps/domain/decorators.py]: @czue 
I checked this and when I try to go to the page from whitelist, I getting 403 error. I tried this for all users, this working as previously but when we enable restriction for stuff users then pages from whitelist are available for these users.

**Comment by lwyszomi** [comment]
@czue 
Updated

File: corehq/apps/domain/decorators.py (hunk 7 above)
Comment (czue) [inline corehq/apps/domain/decorators.py]: hmm, i see I was confused about what this was doing (this is the "no permission" check). why was this necessary? it seems like this is just more restrictive than before (i made this truth table because i was having a hard time wrapping my head around this).

image

**Comment by czue** [comment]
fyi not mergeable anymore. my question isn't blocking but i'm curious the answer

**Comment by lwyszomi** [comment]
@czue 
I updated the PR, You're right with conditions. I extended condition and now we also check superuser flag.
